clang/lib/Tooling/DependencyScanning/IncludeTreeActionController.cpp

```diff
@@ -545,13 +565,82 @@ IncludeTreeBuilder::finishIncludeTree(CompilerInstance &ScanInstance,
       getCASTreeForFileIncludes(IncludeStack.pop_back_val());
   if (!MainIncludeTree)
     return MainIncludeTree.takeError();
-  auto FileList = cas::IncludeTree::FileList::create(DB, IncludedFiles);
+
+  if (!ScanInstance.getLangOpts().CurrentModule.empty()) {
+    SmallVector<cas::ObjectRef> Modules;
+    auto AddModule = [&](Module *M) -> llvm::Error {
+      Expected<cas::IncludeTree::Module> Mod = getIncludeTreeModule(DB, M);
+      if (!Mod)
+        return Mod.takeError();
+      Modules.push_back(Mod->getRef());
+      return Error::success();
+    };
+    if (Module *M = ScanInstance.getPreprocessor().getCurrentModule()) {
+      if (Error E = AddModule(M))
+        return std::move(E);
+    } else {
+      // When building a TU or PCH, we can have headers files that are part of
+      // both the public and private modules that are included textually. In
+      // that case we need both of those modules.
+      ModuleMap &MMap =
+          ScanInstance.getPreprocessor().getHeaderSearchInfo().getModuleMap();
+      M = MMap.findModule(ScanInstance.getLangOpts().CurrentModule);
+      assert(M && "missing current module?");
+      if (Error E = AddModule(M))
+        return std::move(E);
+      Module *PM =
+          MMap.findModule(ScanInstance.getLangOpts().ModuleName + "_Private");
```

Comment:
It's handled specifically, not just a convention. There are various places in ModuleMap that handle private modules by name.  Technically there are three kinds (Foo_Private, Foo.Private, and FooPrivate), but only Foo_Private is relevant here, because that's the only one that is (a) treated as part of the main module in the function `Module::isForBuilding`, and (b) is not already a submodule that would be included as part of `AddModule(M)` above.